docs/en/development/io_processing.md

```diff
@@ -3,7 +3,51 @@ sidebar_label: Read/Write/Delete Processing Flow
 
-Working in progress...
+## Workflow of processing write
 
+JuiceFS splits large files at multiple levels to improve I/O performance. See [how JuiceFS store files](../reference/how_juicefs_store_files.md). When processing write requests, JuiceFS first writes data into the client buffer and stores in chunks and slices. Chunks are isolated logical units each split into 64 MiB offset by offset consequently, and will be further split into slices if necessary. A new slice will be created if it doesn't overlap nor adjoin any existing slices, otherwise, update the affected existing slices in-place instead of creating. Slices are logical units for data persistency. When doing a flush, a slice will be first split into 4 MiB blocks consequently, and then uploaded to object storage, and eventually the metadata of the slice will be upadted. There is a one-to-one correspondence between blocks and objects. As we can see, when writing consequently, **only one** increasing slice and one final flush are needed. This maximizes the write performance of the object storage. A simple [JuiceFS benchmark](../benchmark/performance_evaluation_guide.md) as shows below consequently writes a 1 GiB file at its first stage with a 1 MiB I/O size. The following figure shows the dataflow in each component of the system.
+
+![write](../images/internals-write.png)
+
+> **NOTICE**: Compression and encryption in this figure are disabled by default. To enable these features, please add option `--compress value` or `--encrypt-rsa-key value` to the formatting command.
+
+To be more intuitive, the folloing metrics figure shows the metrics from the `stats` command.
+
+![stats](../images/internals-stats.png)
+
+From the first highlighted section in the above figure, we can see,
+
+- The average I/O size writing to the object storage is `object.put / object.put_c = 4 MiB`. It is the same as the default size of a block.
+- The number of metadata transactions to that of object storage transactions is `meta.txn : object.put_c -= 1 : 16`. It means a single slice flush requires 1 update of metadata and 16 uploads to the object storage. It also shows that every flush operation transmits 4 MiB * 16 = 64 MiB of data as the same to the default size of a chunk.
+- The average request size in the FUSE layer approximately equals to `fuse.write / fuse.ops ~= 128 KiB`, which is the same as the default limitation of the request size.
+
+It's more complicated to write a large file randomly than to write it consequently. There could be **a number of intermittent** slices in a chunk, which makes it hard to crop the data into a size of 4 MiB, and lowers the performance since it requires multiple metadata updates.
+
+Generally, a small file will be uploaded to the object storage when it is being closed, and the I/O size is the same as the file size. It could also be seen from the third highlighted section (creating 128 KiB small files) in the metrics figure above that,
```

Comment:
```suggestion
Generally, when writing a small file, the file will be uploaded to the object storage when it is being closed, and the I/O size is the same as the file size. It could also be seen from the third highlighted section (creating 128 KiB small files) in the above metrics figure that,
```